### PR TITLE
Add test for tidb/issues/49241

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: ref #6233, close https://github.com/pingcap/tidb/issues/49241, 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
/hold

till https://github.com/pingcap/tidb/issues/49241 merged

File: tests/fullstack-test/mpp/apply_with_late_materialization.test [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
++------+
+|    5 |
++------+
+mysql> set @@tidb_allow_mpp=1; set @@tidb_opt_enable_late_materialization=0; select (select t.value from t where t.id = t1.id order by t.value limit 1) xx from t1 order by t1.value limit 5;
++------+
+| xx   |
++------+
+|    5 |
++------+
+mysql> set @@tidb_allow_mpp=0; set @@tidb_opt_enable_late_materialization=1; select (select t.value from t where t.id = t1.id order by t.value limit 1) xx from t1 order by t1.value limit 5;
```
Comment (windtalker) [inline tests/fullstack-test/mpp/apply_with_late_materialization.test:47]: I think you need to set tidb_allow_tiflash_cop to true if tidb_allow_mpp is set to false.

Comment (windtalker) [reply tests/fullstack-test/mpp/apply_with_late_materialization.test:47]: And also add a test for batch cop? You can use `tidb_allow_batch_cop` enforce/disable batch cop

Comment (Lloyd-Pottiger) [reply tests/fullstack-test/mpp/apply_with_late_materialization.test]: done

File: tests/fullstack-test/mpp/apply_with_late_materialization.test [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
++------+
+mysql> set @@tidb_allow_mpp=1; set @@tidb_opt_enable_late_materialization=0; select (select t.value from t where t.id = t1.id order by t.value limit 1) xx from t1 order by t1.value limit 5;
++------+
+| xx   |
++------+
+|    5 |
++------+
+
+# BatchCop
+mysql> set @@tidb_allow_mpp=0; set @@tidb_allow_tiflash_cop = 1; set @@tidb_allow_batch_cop = 1; set @@tidb_opt_enable_late_materialization=1; select (select t.value from t where t.id = t1.id order by t.value limit 1) xx from t1 order by t1.value limit 5;
```
Comment (windtalker) [inline tests/fullstack-test/mpp/apply_with_late_materialization.test:52]: `tidb_allow_batch_cop` can be set to 2, which mean always use batch cop

Comment (Lloyd-Pottiger) [reply tests/fullstack-test/mpp/apply_with_late_materialization.test]: done

**Review by windtalker — APPROVED** [review]
LGTM

File: tests/fullstack-test/mpp/apply_with_late_materialization.test [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+mysql> insert into test.t1 values(2,5);
+# insert more than 8192 rows to make sure late materialization is enabled.
+mysql> insert into test.t select * from test.t; insert into test.t select * from test.t; insert into test.t select * from test.t; insert into test.t select * from test.t;
+mysql> insert into test.t select * from test.t; insert into test.t select * from test.t; insert into test.t select * from test.t; insert into test.t select * from test.t;
+mysql> insert into test.t select * from test.t; insert into test.t select * from test.t; insert into test.t select * from test.t;
+mysql> alter table test.t set tiflash replica 1;
+mysql> alter table test.t1 set tiflash replica 1;
+mysql> analyze table test.t;
+
+func> wait_table test t t
```
Comment (JaySon-Huang) [inline tests/fullstack-test/mpp/apply_with_late_materialization.test:32]:
```suggestion
```

**Comment by JaySon-Huang** [comment]
/hold

The tidb testing image does not contain the necessary commit
```
[2023/12/14 08:49:11.776 +00:00] [INFO] [printer.go:47] ["Welcome to TiDB."] ["Release Version"=v7.6.0-alpha] [Edition=Community] ["Git Commit Hash"=9633ca1fb4993c727dcdcc015d89a2afdea2e4bb] ["Git Branch"=heads/refs/tags/v7.6.0-alpha] ["UTC Build Time"="2023-12-13 11:44:52"] [GoVersion=go1.21.5] ["Race Enabled"=false] ["Check Table Before Drop"=false]
```

**Pull request merged**